### ISSUE TITLE
Analyze DNS link failures

### ISSUE THREAD

**Issue by OliverWang13** [privacy-tech-lab/gpc-web-crawler#57]
I am opening this issue to track the DNS link failures found in the validation crawls performed by @katehausladen.

False positives:

- instacart.com
- farmersalmanac.com
- mathway.com
- alltrails.com
- mashed.com

False negatives:

- manualslib.com
- mgmresorts.com
- sallybeauty.com
- mattressfirm.com
- beliefnet.com

I will be visiting these sites one by one and trying to find the possible causes.

#### instacart.com

When I visit this site I see a "Your Privacy Choices" link at the bottom, so this may not be a false positive.

#### farmersalmanac.com

I see no DNS link on the home page so this is a true false positive. When looking at the network requests, I can see where our regex is being triggered. In a file by Adthrive (one of their ad managers, presumably), there is the following code:
<img width="517" alt="Screen Shot 2023-07-11 at 3 26 30 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/82842273/c4d6d1a7-082b-4a49-8e14-10bf68fe627d">
Interestingly, the usprivacy cookie is 1NYN, even without sending the GPC signal and only using the LA VPN. With our current method of detecting DNS links, I am not sure if this can be resolved. I can try to look further into restricting the net requests we analyze to just the first party.

#### mathway.com

This site has a similar issue to farmersalmanac.com. A do not sell link does exist, but due to the nature of the site it is not really on the home page. The site immediately brings you to a calculator screen, but you can navigate to the DNS link by going [here](https://www.mathway.com/about). This time, the source is cdn.cookielaw.org for the file that triggers our regex.

#### alltrails.com

Similar to the two above, the consent management platform Osano has a file that triggers our regex and causes the false positive.

#### mashed.com

This is also coming from cdn.cookielaw.org just like mathway.com and similarly is incorrectly found.

#### manualslib.com

I do not see a DNS link on this site.

#### mgmresorts.com

This site has a somewhat unique phrasing of "Do not sell/share my personal information" that we can expand the regex to fit.

#### sallybeauty.com

I don't see any reason why we should not pick this up. There is a "your privacy choices" link that is pretty obvious and can be found in the first-party network traffic.

#### mattressfirm.com

Same as above, unsure why we aren't picking it up. The link is also found in the first-party network traffic.

#### beliefnet.com

Same as the two above. There's a link containing regex we should be able to identify and it can be found in the first-party network traffic.

### Conclusions

It is clear to see why most of the false positives occurred and I do not think there is a ton that can be done differently to avoid these, but I can take a look at our detection method.

Most of the false negatives are a bit more confusing and might require some further investigation, although it will be helpful to look over the detection code again in order to look for a possible cause.

**Comment by katehausladen** [comment]
I also see Your Privacy Choices on instacart.com. I guess I missed it before.

**Comment by SebastianZimmeck** [comment]
> It is clear to see why most of the false positives occurred and I do not think there is a ton that can be done differently to avoid these, but I can take a look at our detection method.

> Most of the false negatives are a bit more confusing and might require some further investigation, although it will be helpful to look over the detection code again in order to look for a possible cause.

@katehausladen, if you could also investigate the false positives and negatives, that would be great. Maybe, you guys together can come up with ideas what we could do to fix these.

**Comment by katehausladen** [comment]
for manualslib.com, this is the DNSL I saw.
<img width="600" alt="Screenshot 2023-07-20 at 10 25 49 AM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/6634d7ed-3cf5-49b2-bf43-7a1129a0d404">
and this is the code for it in the html
<img width="600" alt="Screenshot 2023-07-20 at 10 32 17 AM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/4b8457a8-9736-4273-94e3-cfb2005f4302">
It's set with this file
<img width="600" alt="Screenshot 2023-07-20 at 10 31 19 AM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/dfaa3f25-8d46-46dc-beff-8bee6cf9e8a5">

**Comment by katehausladen** [comment]
I crawled again today after adjusting some timeouts. From today's crawl:

false negatives:
- manualslib.com
- michaelkors.com
- mattressfirm.com
- beliefnet.com

false positives:
- farmersalmanac.com
- mathway.com
- alltrails.com
- mashed.com

All except michaelkors.com were previously incorrect, but we did get a few more of the ones that failed in the previous crawl. After visiting michaelkors.com, I saw their DNSL is now "California Consumer DNS"

<img width="700" alt="michaelkors_dnsl" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/0b6dac0e-16fb-4524-add9-16022942860c">

When I created the ground truths file on 7/5, I noted that this site's DNSL was "California Residents: Do not sell my information." 

Is this new link something we should add to the regex? Or is this phrasing technically not compliant?

**Comment by SebastianZimmeck** [comment]
> Is this new link something we should add to the regex? Or is this phrasing technically not compliant?

Technically, it is not compliant. Though, I would still add it because they are doing something, albeit, in an incorrect way.

**Comment by OliverWang13** [comment]
I think I would lean toward not adding it to the regex, but if you think it's important to have I can add it.

**Comment by SebastianZimmeck** [comment]
> I think I would lean toward not adding it to the regex

What are the reasons? Such link is not compliant, would complicate the regex too much, too many false positives, ...?

**Comment by OliverWang13** [comment]
Because the link is not compliant in the eyes of the law I think that the site may not technically have a "do not sell link". If we were to try and use Optmeowt as a legal tool of any sort it would not make sense to include this phrasing. It would complicate the regex but that is not make or break for me.

**Comment by katehausladen** [comment]
For consistency purposes, I think we should come up with a concrete protocol for determining ground truth values for all the variables we collect. More specifically, I think we need to decide how to deal with edge cases. Maybe for DNSLs, our concrete protocol is that our regex list will contain all CCPA-compliant phrasing, and we say DNSL = True only if the link on the page would match our regex.

But, how do we classify Dropbox? Dropbox has a link called "Cookies and CCPA Preferences" at the bottom of their site.
<img width="600" alt="Screenshot 2023-07-25 at 12 45 15 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/02769855-b512-473f-9a59-18531abe597a">

While it may not match our regex, it is clearly aimed at the CCPA and opens a popup window (without leaving the home page) that does have the proper phrasing. Since this popup window is still part of the home page, our regex does pick it up. Would we consider this site not compliant? 
<img width="600" alt="Screenshot 2023-07-25 at 12 45 22 PM" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/adf95a4d-fc3c-4410-8e35-610621c96ab9">

I also found a site with a weird USPAPI implementation. Popsugar replies twice to every call to the USPAPI. While Popsugar's implementation is clearly wrong, we can't really call it incorrect if our extension replies with "1YNY" or "null." 
<img width="600" alt="popsugar" src="https://github.com/privacy-tech-lab/gpc-web-crawler/assets/40359590/75035085-cc7c-42f3-8842-95b1390f9fd4">
What would we say is the ground truth USPAPI value for this site?

**Comment by katehausladen** [comment]
> Beyond that, we probably would not see such instance in our dataset as we do not check the groundtruth, right?

That's true, and since the USPAPI is being deprecated, this shouldn't be an issue going forward.

Since we're shifting our focus to issue #59, I'll close this issue for now.
